File: Homework/Hoang_Nguyen/lesson3/622_designCircularQueue.js
/**
 * @param {number} k
 */
var MyCircularQueue = function (k) {
  this.arr = [];
  this.size = k;
  this.head = -1; // empty queue
  this.tail = -1;
};

/**
 * @param {number} value
 * @return {boolean}
 */
MyCircularQueue.prototype.enQueue = function (value) {
  if (this.isFull()) return false;
  if (this.isEmpty()) {
    this.head++;
  }
  this.tail = (this.tail + 1) % this.size;
  this.arr[this.tail] = value;
  return true;
};

/**
 * @return {boolean}
 */
MyCircularQueue.prototype.deQueue = function () {
  if (this.isEmpty()) return false;
  // queue has 1 element -> set to empty
  if (this.head === this.tail) {
    this.head = -1;
    this.tail = -1;
  } else {
    this.head = (this.head + 1) % this.size;
  }
  return true;
};

/**
 * @return {number}
 */
MyCircularQueue.prototype.Front = function () {
  if (this.isEmpty()) return -1;
  return this.arr[this.head];
};

/**
 * @return {number}
 */
MyCircularQueue.prototype.Rear = function () {
  if (this.isEmpty()) return -1;
  return this.arr[this.tail];
};

/**
 * @return {boolean}
 */
MyCircularQueue.prototype.isEmpty = function () {
  return this.head === -1;
};

/**
 * @return {boolean}
 */
MyCircularQueue.prototype.isFull = function () {
  return (this.tail + 1) % this.size === this.head;
};

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * var obj = new MyCircularQueue(k)
 * var param_1 = obj.enQueue(value)
 * var param_2 = obj.deQueue()
 * var param_3 = obj.Front()
 * var param_4 = obj.Rear()
 * var param_5 = obj.isEmpty()
 * var param_6 = obj.isFull()
 */

File: Homework/quoc_khanh/lesson_2/622.DesignCircularQueue.py
class MyCircularQueue:

    def __init__(self, k: int):
        self.array = [0] * k
        self.length = 0
        self.capacity = k
        self.rear = -1
        self.front = 0

    def enQueue(self, value: int) -> bool:
        '''Insert at rear'''
        if self.isFull():
            return False
        else:
            self.rear = (self.rear + 1) % self.capacity
            self.array[self.rear] = value
            self.length += 1
            return True

    def deQueue(self) -> bool:
        '''Delete at front'''
        if self.isEmpty():
            return False
        else:
            self.front = (self.front + 1) % self.capacity
            self.length -= 1
            return True
        

    def Front(self) -> int:
        if self.isEmpty():
            return -1
        else:
            return self.array[self.front]

    def Rear(self) -> int:
        if self.isEmpty():
            return -1
        else:
            return self.array[self.rear]

    def isEmpty(self) -> bool:
        return self.length == 0

    def isFull(self) -> bool:
        return self.length == self.capacity

s = MyCircularQueue(3)
print(s.enQueue(1) == True)
print(s.enQueue(2) == True)
print(s.enQueue(3) == True)
print(s.enQueue(4) == True)
print(s.Rear() == 3)
print(s.Rear())